src/managedwriter/stream_connection.ts

```diff
@@ -106,62 +117,49 @@ export class StreamConnection extends EventEmitter {
 
   private handleError = (err: gax.GoogleError) => {
     this.trace('on error', err, JSON.stringify(err));
-    if (this.shouldReconnect(err)) {
-      this.reconnect();
-      return;
-    }
-    let nextPendingWrite = this.getNextPendingWrite();
-    if (this.isPermanentError(err)) {
-      this.trace('found permanent error', err);
-      while (nextPendingWrite) {
-        this.ackNextPendingWrite(err);
-        nextPendingWrite = this.getNextPendingWrite();
-      }
-      this.emit('error', err);
-      return;
-    }
-    if (this.isRequestError(err) && nextPendingWrite) {
+    const nextPendingWrite = this.getNextPendingWrite();
+    if (nextPendingWrite) {
       this.trace(
         'found request error with pending write',
         err,
         nextPendingWrite
       );
-      this.ackNextPendingWrite(err);
+      this.handleRetry(err);
+    }
+    if (this.isRetryableError(err) && this.listenerCount('error') === 0) {
       return;
     }
     this.emit('error', err);
   };
 
-  private shouldReconnect(err: gax.GoogleError): boolean {
-    const reconnectionErrorCodes = [
-      gax.Status.UNAVAILABLE,
-      gax.Status.RESOURCE_EXHAUSTED,
+  private handleRetry(err: gax.GoogleError) {
+    const retrySettings = this._writeClient['_retrySettings'];
+    if (this.isRetryableError(err) && retrySettings.enableWriteRetries) {
+      if (!this.isConnectionClosed()) {
+        const pw = this._pendingWrites.pop()!;
+        this.send(pw);
+      }
```

Comment:
In the rare case that conneciton is closed, we didn't send the request again, then we should make sure we nack the request.

I think we should do
if (retry) {
   if (connection not closed) {
       send;
       return;
    }   
}
nack;

Reply:
the nack is going be handled on the `'close'` callback in that scenario.